### PR TITLE
Fix idempotency issues in set_bios_attributes

### LINKED ISSUE #46338: redfish_config: SetBiosAttributes not idempotent
I haven't found another issue about this. Please forgive me if I open this and it turns out I was just too stupid to search.

##### SUMMARY
I have had a look at the new redfish modules. I know 2.7 isn't released yet and anyway these modules are flagged as preview, but I was somewhat irritated by the behavior of SetBiosAttributes in redfish_config. I've written a task to set the BIOS attribute EnergyPerformanceBias to MaxPower. Although the attribute already has this value, the action is executed and the task therefore says "changed". The other modules I have worked with so far are idempotent, they would discover that the value is already correct and report a status of (already) "OK". Would it be possible to make redfish_config idempotent?

##### ISSUE TYPE
Feature Idea

##### COMPONENT NAME
!component = lib/ansible/modules/remote_management/redfish/redfish_config.py

##### ANSIBLE VERSION
```
ansible 2.7.0rc4
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.13 (default, Oct 26 2017, 01:54:36) [GCC 6.3.0]
```

##### CONFIGURATION
```
```
##### OS / ENVIRONMENT
PowerEdge R740xd, iDRAC Firmware Version 3.15.17.15

##### STEPS TO REPRODUCE
```
- hosts: localhost
  connection: local
  tasks:
    - name: Set EnergyPerformanceBias to MaxPower
      redfish_config:
        category: Systems
        command: SetBiosAttributes
        bios_attr_name: EnergyPerformanceBias
        bios_attr_value: MaxPower
        baseuri: {{ baseuri }}
        user: "{{ user }}"
        password: "{{ password }}"
```

##### EXPECTED RESULTS
I expect that EnergyPerformanceBias isn't set to MaxPower if this is already its value:
```
PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
ok: [localhost]

TASK [Set EnergyPerformanceBias to MaxPower] ***********************************
ok: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0
```
##### ACTUAL RESULTS
```
PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
ok: [localhost]

TASK [Set EnergyPerformanceBias to MaxPower] ***********************************
changed: [localhost]

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0
```

##### ADDITIONAL INFORMATION
We are just starting using ansible for configurtion management. Therefore we still have a close look at the PLAY RECAP after running a playbook. Every task that changed something wants to be investigated, i.e. did we make a mistake in the playbook or was the system really configured wrong and ansible just did its job. When tasks announce changes that effectively didn't change anything, there is a lot of work to figure out what's going on. And anyway: It's such a good feeling when ansible reports everything as OK.

### PR DESCRIPTION
**Pull request by jose-delarosa**
- Added check to see if attribute even exists, if not, it exits.
- Then checks if attribute is already set to value we want to update
  it to. If yes, then it exits and changed=False
- Otherwise updates the attribute and changed=True

##### SUMMARY

Fixes idempotency issues in set_bios_attributes by checking first if the desired BIOS attribute is already set to the desired value.

Fixes https://github.com/ansible/ansible/issues/46338

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

redfish_config.py
redfish_utils.py

##### ANSIBLE VERSION

```paste below
ansible 2.8.0.dev0 (fix-idempotency-issues 7c2ae6f4aa) last updated 2018/10/22 15:36:50 (GMT -500)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /share1/git/ansible/lib/ansible
  executable location = /share1/git/ansible/bin/ansible
  python version = 2.7.5 (default, May 31 2018, 09:41:32) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

##### ADDITIONAL INFORMATION

Below is output of running same command (enable PXE in NIC port 4) twice.
Playbook: https://github.com/dell/redfish-ansible-module/blob/master/playbooks/bios/enable_pxe.yml

```paste below
[root@ans1 bios>ansible-playbook enable_pxe.yml

PLAY [Enable PXE in NICs] *********************************************************************

TASK [Set PxeDev4EnDis to Enabled] ************************************************************
changed: [red3]

TASK [Schedule Config Job - Reboot] ***********************************************************
changed: [red3]

PLAY RECAP ************************************************************************************
red3                       : ok=2    changed=2    unreachable=0    failed=0    skipped=0

Playbook run took 0 days, 0 hours, 0 minutes, 7 seconds
[root@ans1 bios>ansible-playbook enable_pxe.yml

PLAY [Enable PXE in NICs] *********************************************************************

TASK [Set PxeDev4EnDis to Enabled] ************************************************************
ok: [red3]

TASK [Schedule Config Job - Reboot] ***********************************************************
skipping: [red3]

PLAY RECAP ************************************************************************************
red3                       : ok=1    changed=0    unreachable=0    failed=0    skipped=1

Playbook run took 0 days, 0 hours, 0 minutes, 2 seconds

```

### CODE REVIEW

File: lib/ansible/modules/remote_management/redfish/redfish_command.py
Comment (billdodd) [inline lib/ansible/modules/remote_management/redfish/redfish_command.py:252]: Should this be using the `changed` and `msg` values from `result`? e.g.:
```
module.exit_json(changed=result['changed'], msg=to_native(result['msg']))
```

**Review by billdodd — COMMENTED** [review]
I reviewed the code. Looked good, but I did have one question in redfish_command.py. See my comment inline in that file.

**Comment by jose-delarosa** [comment]
So in redfish_command the approach is different in that it might be harder to check the state of a system before a command is executed. It's possible in some instances, but needs to be checked for each function. I concluded that the best thing for now was to assert that the command was executed on the system **regardless**, and so changed=True needs to be passed back. We can certainly take a look here, but that would be a separate issue, a separate PR.

**Comment by jose-delarosa** [comment]
By the way that line that I removed in redfish_command.py was sloppy committing on my side, it's OK to remove since its not being used, but should have been done in a separate commit and perhaps in a separate PR since it has nothing to do with the issue that this PR addresses.

**Comment by billdodd** [comment]
Jose - thanks for the reply. That makes sense.

**Review by billdodd — APPROVED** [review]
Changes look good. I'm guessing the modules_utils updates still need manual approval for merging, but I'll still go ahead and add:

shipit

**Comment by jose-delarosa** [comment]
Not sure if redfish_utils will merge automatically, there's a PR for that, not sure if issue is resolved. We'll find out.

shipit

**Comment by jose-delarosa** [comment]
If it does not merge automatically, need to bring it up to the core team. They meet today at 2 pm CST in IRC, channel ansible-meeting I think. Can add to the agenda https://github.com/ansible/community/issues/375 and bring up. I am away from my IRC client so can't make it today. The core team meets again on Thursday, I can bring it up then.

**Pull request merged**
